### ISSUE TITLE
switch-to-configuration-ng: Tries to stop transient mount units

### ISSUE THREAD

**Issue by Mic92** [NixOS/nixpkgs#343975]
### Describe the bug

Discovered in https://github.com/NixOS/nixpkgs/pull/343709
it seems that the new switch-to-configuration-ng will try to stop mount units that are not defined by NixOS. This is bad because it can break all kinds of desktop automount, nix-store mounts
or mounts generated by other services i.e. docker or zfs snapshots

### Steps To Reproduce

```
$ nix-build ./nixos/release.nix -A tests.nix-upgrade 
```

### Expected behavior

switch-to-configuration-ng should respect the IgnoreOnIsolate or Transient property of a unit and not stop those.

### Metadata
Please run `nix-shell -p nix-info --run "nix-info -m"` and paste the result.

```console
[user@system:~]$ nix-shell -p nix-info --run "nix-info -m"
output here
```

---

Add a :+1: [reaction] to [issues you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[issues you find important]: https://github.com/NixOS/nixpkgs/issues?q=is%3Aissue+is%3Aopen+sort%3Areactions-%2B1-desc

**Comment by Mic92** [comment]
cc @NixOS/systemd

**Comment by arianvp** [comment]
I think any transient unit should be ignored.

E.g. Gnome creates `.scope` units for each application.   `systemd-run` creates `.service` units for each invocation.  These are things we don't want to stop during `stc`

**Comment by Mic92** [comment]
Can't tell if anything but mount units are affected by this.

**Comment by ElvishJerricco** [comment]
Why isn't this a problem with the non-ng stc?

**Comment by ElvishJerricco** [comment]
> E.g. Gnome creates .scope units for each application.

@arianvp I don't think these are relevant since they're user units. I think a lot of the stc logic would be awful for gnome if they affected user units.

**Comment by Mic92** [comment]
So for the nix-store.mount it must be generated by the systemd-mount generator because nothing in nix will generate this unit. So I assume that also other mounts would be affected by this i.e. those create by gnome's disk mount.

**Comment by ElvishJerricco** [comment]
I see `Failed unmounting /nix/store.` with perl stc as well. I wonder if the difference isn't what they're doing; but rather that ng is exiting with a fail code where perl wouldn't.

**Comment by ElvishJerricco** [comment]
I guess the next question is... why is this only happening in this particular test? Isn't this mount unit always transient like this on NixOS, because of the ro bind mount?

**Comment by jmbaur** [comment]
For the nix-upgrade test in particular, I created this PR to address the failing test: https://github.com/NixOS/nixpkgs/pull/346702

**Comment by ElvishJerricco** [comment]
After some investigation, I'm fairly sure that the cause of the test failure that prompted this issue was not a difference in the actual *behavior* of stc-ng, but just a difference in the exit code. With stc-ng, it's receiving the dbus signal from systemd that the stop job failed and reporting that with an exit code. stc-perl doesn't notice the failure, because it just iterates over active units at the end looking for failures. But in the case of these mount units failing to stop, they remain mounted and therefore active. So the failed stop happens both ways, but stc-ng gets the fail signal, and stc-perl doesn't even notice.

If we *really* want perfect bug-for-bug compatibility with stc-perl, then we can add code to the `finished_jobs` processing that checks if a failed unit is still failing and otherwise ignores it with a warning, and add a comment saying "This is bad behavior and we should remove it post 24.11 when we aren't striving for perfect bug-for-bug compatibility". But frankly, I think the difference between stc-ng and stc-perl here is negligible, and I think the way stc-ng does it is better. So I'm tempted to say we leave it.

Now, that doesn't mean we should close this issue. The fact that both stc-ng and stc-perl try to stop the transient mount unit is itself arguably a bug. I'm just saying the difference between stc-ng and stc-perl in this case doesn't seem to matter.

**Comment by ElvishJerricco** [comment]
Actually, it doesn't even try to stop transient mount units. It only stops units that have been removed from `/etc/systemd/system/` or from `fstab`. So the reason the original issue happened was because `/nix/store` had been removed from `fstab`, which is (kind of) technically correct behavior (though I can see we have stuff to avoid doing this to `/` and `/nix`, but not `/nix/store`).

**Comment by Mic92** [comment]
We were able to fix at least the failing nix-upgrade test.

**Comment by ElvishJerricco** [comment]
@Mic92 Yes, because the test was switching from a configuration that had a `/nix/store` mount in `fstab` to one that didn't. The fix was to switch instead to a configuration that doesn't remove that mount from `fstab`. With stc-perl, we had exactly the same problem; it tried to stop `nix-store.mount`. But it wasn't reported as an error because the script was insufficient to realize there was an error.

**Comment by jmbaur** [comment]
I'm happy to keep the current behavior, and I doubt many are depending on the current lack of failure of stc-perl for changing mounts (of course besides this test, prior to the fix).
